### ISSUE TITLE
Tips to remove stutter

### ISSUE THREAD

**Issue by MrPhaze62** [MrPhaze62/moonlight-chrome-tizen-no-gamemode#1]
If your stream is stuttery when gaming or watching a youtube video with this fork that disables gamemode at launch, then follow the steps below on your samsung tvs that has these options.

Reason why the stutters are noticable is because of video enhancement made by samsung TV settings, when gamemode is enabled on Kyro and OneLiberty forks, all video enhancement made by samsung is disabled thus making it a smooth stream on their forks. This is fine for many users. But means most samsung settings are disabled like picture brightness/enhancement.

In my case, the original forks that has gamemode enabled on launch, is too bright for my eyes and cannot lower the brightness as they were disabled on launch amonst other settings.

This fork however has disabled gamemode on launch meaning video enhancement is enabled as if you're watching normal TV, which is fine for normal TV, not so fine for Moonlight. But you are able to configure picture brightness and so on and options are not disabled.

**Tips to remove stutters on your samsung TV with this port that has no gamemode enabled**

On your TV:

1) disable picture/video enhancement setting like judder effect and others like so, it may be called picture clarity on some newer TVs. Switch it to off on samsung tvs.

2) disable constrast enhancement settings.

Moonlight stream should now be smooth.

However, if you exit the app, you should remember to change the settings you've made back to normal if you are watching normal TV.

Also, the colour buttons on the TV remote may not work during stream due to gamemode being disabled.

**Comment by CyberShivago** [comment]
Would it be too much trouble to ask what is the line of code that forces the game mode on and off? There is another fork which has multiple additional enhancements like 120,144 FPS and more. It's also very smooth with zero stuttering issues unlike oneliberty. However, it does not force game mode and I actually want game mode due to the removal of additional lag from the controllers. 

Would you mind telling me what is the line of code that forces game mode to be on and off?

**Comment by MrPhaze62** [comment]
@CyberShivago Apologies for the late reply, I rarely get github notifications as I am usually active over on discord.

But to answer your question, The line you're looking for is found in Main.cpp which is

"samsung::wasm::ElementaryMediaStreamSource::LatencyMode::kUltraLow" this is found in forks that has gamemode enabled and is the line that enables gamemode ON, keeping the latency at its lowest possible set by samsung.

In my case to disable gamemode entirely, I changed the word "KUltraLow" into "KLow" to disable gamemode and retaining picture options and such at the cost of latency.

This is also found in samsung documentations found here: https://developer.samsung.com/onlinedocs/tv/tizen_wasm_player/classsamsung_1_1wasm_1_1_elementary_media_stream_source.html#af2a038fbc524b7a50a71c6d32847530c


You can see the exact commit I've made awhile back for this change here: https://github.com/MrPhaze62/moonlight-chrome-tizen-no-gamemode/commit/bb9f1746c2f65919fadb76c7c05671815e055b83

**Comment by CyberShivago** [comment]
I took toypoodlegamings fork and left it exactly the same but recompiled it with KUltraLow, And I can say with 100% certainty that when you enable game mode you also get micro stuttering. It's not frame per second loss but simply frame sync problems.

It's really unfortunate though, I wish game mode worked without creating this consistent stutter. If we could somehow figure out what causes this, The code would be flawless for me. Well I would say for anyone that wants to use game mode. I know that there isn't a significant enough lag to be concerned, But I went back and forth between moonlight installations and I would say that yes there is more lag without game mode on. I actually think that the lag has something to do with the controller to the television communication. Like last Bluetooth lag? I don't know really. 

But I do know this. Game mode creates micro stutters with moonlight sadly, so until some genius solves the problem I can't use it. I'm actually quite surprised that no one else has noticed all forks that have game mode enabled have a stutter and all forks that do not have no stutter.

**Comment by Desoulen** [comment]
I've noticed the same. All forks with game mode on I'll get the original HDR stream and low latency. Unfortunately all of them have stutter and audio cracks. The only fork which works for me well is this one. So far amazing work and I really enjoy with super low decoding time to stream my games without the need of a minipc whatsoever. I'm really hoping the game mode stutter on tizen 9 will be fixed as @CyberShivago said. The lower latency (most prob Controller as all ms for Network and decoding etc are super low) is really noticable when playing Hades 2 e.g. 

Anyone know what this could potentially cause? Is it a Tizen issue?

**Comment by MrPhaze62** [comment]
@Desoulen Yeah, the controller input lag/issue has been reported across multiple forks and in discord, it's likely a Tizen issue as a whole, but as of now there isn't any fix for it, originally I thought it would be a network issue for others as I personally have not experienced input lag delay, but seeing more reports about it makes it more likely now that there is indeed some delay. On my end though it seems manageable or perfect for me. I do not know if it is just specific samsung tvs or just a Tizen issue as a whole, even though the video and audio are perfectly fine. I wish I could look into the issue but I'm not experienced enough for this, and I'm sure other forks of moonlight tizen would love to figure this out as well if it also happens on their forks.

**Comment by CyberShivago** [comment]
I'd like to follow up with this conversation to say that I have used a new fork thst I believe solves this issue. 

https://github.com/ruanformigoni/moonlight-tizen

Now I need to give some details just in case it doesn't work for anyone else. I have seen it cause the micro stutter as well, But there is a settings menu that's built into this version that I've never seen on Tizen Moonlight before. The reason I mentioning it is because when I did see the micro stutter I turned the HDR off and back on again and when reconnecting the micro stutter disappears. 

I'm not exactly sure what is happening here but I think that it has something to do with catching the correct synchronicity. And the reason why I tried it out is because of this list: 

Overview of the changes:

Replaced the audio backend from EMSS to OpenAL.
Feed OpenAL with a dedicated feeder thread to reduce issues.
Fixed speedup/slowdows which happened randomly throughout the stream.
Fixed the application muting when app switching.
Greatly reduced audio crackling, it is not noticiable anymore on my 5GHz wifi.
Greatly reduced video stuttering by decopling the audio and video processing pipelines.

Now I can use it at 4K and 60Mbps smoothly with low latency selected (Which is also an option in the menu now). 

All these forks have been incredible but only this one gives the ability to change these settings on the fly without having to reinstall a different fork. I don't understand naming conventions very well so I guess it's called "ruanformigoni" fork?

It seems that they recognize and thanked a lot of the other forks including this one for their work along the way which resulted their version of Moonlight.

**Comment by Desoulen** [comment]
I've tried this one as well and I'm stuck on the loading screen (I'm not the only one) and doesn't establish any connection. The host gets the command and launches the game but the fork does not establish a finished connection. I need to wait for some fixes most probably. OpenAl gives me way too much delay in the Audio when I use previous releases. Only the newest with web audio doesn't work for me. But thank you for the tip!

**Comment by MrPhaze62** [comment]
As far as I know ruanformigoni fork of moonlight is very much new and still a work in progress and will soon later if merged with Brightcraft fork of moonlight once fully developed and tested. Early days but ideally that fork is recommended for people having audio stutters and such as a test, but a welcome step forward as people had complained about audio issues and stutters which that fork is an impressive feat.

For this particular issue here, it may not be the fix needed. But maybe in the future who knows, in the end, what works for you, just simply works. My no-gamemode build is based on OneLiberty moonlight fork but with just gamemode entirely disabled, which anyone can do.

Ideally if anyone does want an updated latest build that is actively maintained, that would be brightcraft moonlight-tizen fork and may contain future fixes that may get resolved that this build may be behind in.

In the end, if mine or other forks works for you, then it works and just stick with it until you feel like changing it with another build.

Otherwise, there is a way better method preferred by the official moonlight devs that are maintained by them, which is to use a  firestick/appleTV/android TV ect... plugged to the TV instead with moonlight installed, and that would resolve many of the issues here and others have mentioned natively for samsung tvs.

Each have their own ways and convenience.

**Comment by Desoulen** [comment]
That's absolutely (almost) true! So far I'm using your fork and I'm happy how it runs. Just not working HDR (because of game mode?) and the input lag compared to other forks are the main reasons why it's not 110% perfect (yet).

Unfortunately my Amazon Stick, freshly unpacked, updated 4k max can't handle Moonlight. 10-20ms decoding time plus alot of stutter compared to your butter smooth fork running native on the Samsung itself. Maybe I'm just unlucky and all my devices suck when chained together. :D

**Comment by MrPhaze62** [comment]
I think you can get hdr to work, you have to kind of switch hdr off and then on within windows settings, I think this was mentioned in the main OneLiberty github issues in the past as a bug, but I don't remember. Mind you I do not have a hdr compatible monitor to display hdr so I don't know what it (hdr) looks like even to the TV. I do remember someone saying it did look blown out in the main fork until they turned it off and on again within windows while streaming.

I don't know if it is the case here.

It could also be a side effect with gamemode disabled here as well. That is something i would like to test if I indeed had a HDR compatible monitor.

**Comment by Desoulen** [comment]
@MrPhaze62 if you meant your gaming monitor does not support HDR: https://github.com/VirtualDrivers/Virtual-Display-Driver

Just install oneclick, it fakes a 4k HDR Monitor. In Sunshine make sure to toggle "activate as main monitor" or sth. In the config paste the monitor ID (found in troubleshooting CLI in Sunshine settings, scroll down and find an ID like: {xxxxxxxx...} for the fake monitor. Make sure under windows settings to set up the fake monitor,  activate HDR, refresh rate, resolution.

Your TV should then ofc should support HDR but with other forks I'll get a crazy nice colored picture. Mandatory to calibrate your TV properly with good Film/Game presets.
